flax/linen/recurrent.py

```diff
@@ -726,12 +759,143 @@ def _select_last(sequence: A, segmentation_mask: jnp.ndarray, axis: int) -> A:
   last_idx = segmentation_mask.sum(axis=-1) - 1
 
   def _slice_array(x: jnp.ndarray):
-    _last_idx = _expand_dims_for(last_idx, to_target=x)
+    _last_idx = _expand_dims_like(last_idx, target=x)
     x = jnp.take_along_axis(x, _last_idx, axis=axis)
     return x.squeeze(axis=axis)
 
   return jax.tree_map(_slice_array, sequence)
 
-def _expand_dims_for(x, *, to_target):
+def _expand_dims_like(x, *, target):
   """Expands the shape of 'x' to match those of 'to_target' by adding singleton dimensions."""
-  return x.reshape(x.shape + (1,) * (to_target.ndim - x.ndim))
\ No newline at end of file
+  return x.reshape(x.shape + (1,) * (target.ndim - x.ndim))
+
+# TODO: Make flip_sequences a method of RNN and generalize it to work with
+# multiple batch dimensions.
+def flip_sequences(
+  inputs: Array, segmentation_mask: Optional[Array], num_batch_dims: int, time_major: bool
+) -> Array:
+  """Flips a sequence of inputs along the time axis.
+
+  This function can be used to prepare inputs for the reverse direction of a
+  bidirectional LSTM. It solves the issue that, when naively flipping multiple
+  padded sequences stored in a matrix, the first elements would be padding
+  values for those sequences that were padded. This function keeps the padding
+  at the end, while flipping the rest of the elements.
+
+  Example:
+  ```python
+  inputs = [[1, 0, 0],
+            [2, 3, 0]
+            [4, 5, 6]]
+  lengths = [1, 2, 3]
+  flip_sequences(inputs, lengths) = [[1, 0, 0],
+                                     [3, 2, 0],
+                                     [6, 5, 4]]
+  ```
+
+  Args:
+    inputs: An array of input IDs <int>[batch_size, seq_length].
+    lengths: The length of each sequence <int>[batch_size].
+
+  Returns:
+    An ndarray with the flipped inputs.
+  """
+  # Compute the indices to put the inputs in flipped order as per above example.
+  time_axis = 0 if time_major else num_batch_dims
+  max_steps = inputs.shape[time_axis]
+
+  if segmentation_mask is None:
+    # reverse inputs and return
+    inputs = jnp.flip(inputs, axis=time_axis)
+    return inputs
+
+  lengths = jnp.sum(segmentation_mask, axis=time_axis, keepdims=True) # [*batch, 1]
+  # create indexes
+  idxs = jnp.arange(max_steps - 1, -1, -1) # [max_steps]
+  if time_major:
+    idxs = jnp.reshape(idxs, [max_steps] + [1] * num_batch_dims)
```

Comment:
Nit: Above you use `(1,)*dims` (inside `_expand_dims_like`), while here you use `[1]*dims` instead. I suggest using either that one or this one everywhere consistently.

Reply:
Changed everything to use list concatenations.